### ISSUE TITLE
Linux installation of TauDEM

### ISSUE THREAD

**Issue by solomonvimal** [dtarb/TauDEM#149]
Dr. Tarboton (and TauDEM developers), 

I am getting the following error when I use d8flowdir. I think it has to do with incorrect installation, but pitremove seems to work fine with this installation. I'm using ubuntu 16.04 and TauDEM version 5.0.6. 

```
svimal@lettenmaierlab11:~/Desktop/HAND$ d8flowdir -fel /home/svimal/Desktop/HAND/n10e105_dem_fel.tif -p /home/svimal/Desktop/HAND/n10e105_dem_d8.tifD8FlowDir version 5.0.6
This run may take on the order of 25 minutes to complete.
This estimate is very approximate. 
Run time is highly uncertain as it depends on the complexity of the input data 
and speed and memory of the computer. This estimate is based on our testing on 
a dual quad core Dell Xeon E5405 2.0GHz PC with 16GB RAM.
Error opening file .
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD 
with errorcode 22.
NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
```

Do you know what this error means?  

I managed to install a newer version (5.3.8) of TauDEM in another machine (CentOS 6), but there I could not get dinfdistdown to work, but d8flowdir seems to work fine.

For installation I followed the steps [here](http://docs.qgis.org/2.14/en/docs/user_manual/processing/3rdParty.html#installing-taudem-under-linux) and [here](https://unix.stackexchange.com/questions/346775/how-to-solve-building-error-where-a-variable-was-not-declared-in-this-scope). Is there any other installation guide for Linux that you recommend?

Thanks,
Solomon

**Comment by dtarb** [comment]
Try look in the version of code you are using for MPI_ABORT with an argument of 22.  That will indicate the line it fails (sorry I do not have code for the 5.0.6 version easily accessible to me now).

Compressed files were only enabled by moving to GDAL at version 5.3 so earlier versions will not work with compressed files (but this should give a file read error, and occur with pitremove.

Check grid size and memory demands.  RAM required is 16 times number of cells in bytes.  Compare this to what your computer has, you may have memory overflow problems.

**Comment by solomonvimal** [comment]
Dr. Tarboton, 

Thanks for the suggestions! I tried with a small file. So its surely not a RAM related error. 

Is there any (recent) version of TauDEM that has been tested in Linux (Ubuntu or Debian)? 

I tried to compile the lastest TauDEM (TauDEM-Develop) in Ubuntu 16.04. But I get another error (shown below). Looks like the recent changes to file tiffIO.cpp (mentioned in UpdateNotes.txt) could be the cause. I do not understand the code enough to figure out a clean installation.

I tried older TauDEM versions 5.3.5 (and .4,.3,.2,.1), but they all seem to have the issue reported [here](https://github.com/dtarb/TauDEM/issues/14)

[These](http://docs.qgis.org/2.14/en/docs/user_manual/processing/3rdParty.html#installing-taudem-under-linux) are the steps I follow for installing TauDEM. Would be great to hear if someone has had a successful installation in Ubuntu 16.04. Thanks for your time! 

- Solomon

The error I get is:

`svimal@lettenmaierlab06:~/Software/TauDEM/TauDEM-Develop/src/build$ make
[  1%] Building CXX object CMakeFiles/streamnet.dir/tiffIO.cpp.o
In file included from /home/svimal/Software/TauDEM/TauDEM-Develop/src/commonLib.h:103:0,
                 from /home/svimal/Software/TauDEM/TauDEM-Develop/src/tiffIO.h:49,
                 from /home/svimal/Software/TauDEM/TauDEM-Develop/src/tiffIO.cpp:47:
/home/svimal/Software/TauDEM/TauDEM-Develop/src/linearpart.h: In member function ‘void linearpart<datatype>::savedxdyc(tiffIO&)’:
/home/svimal/Software/TauDEM/TauDEM-Develop/src/linearpart.h:523:16: error: invalid use of incomplete type ‘class tiffIO’
      dxc[i]=obj.getdxc(globalY);
                ^
In file included from /home/svimal/Software/TauDEM/TauDEM-Develop/src/linearpart.h:42:0,
                 from /home/svimal/Software/TauDEM/TauDEM-Develop/src/commonLib.h:103,
                 from /home/svimal/Software/TauDEM/TauDEM-Develop/src/tiffIO.h:49,
                 from /home/svimal/Software/TauDEM/TauDEM-Develop/src/tiffIO.cpp:47:
/home/svimal/Software/TauDEM/TauDEM-Develop/src/partition.h:46:7: note: forward declaration of ‘class tiffIO’
 class tiffIO;
       ^
In file included from /home/svimal/Software/TauDEM/TauDEM-Develop/src/commonLib.h:103:0,
                 from /home/svimal/Software/TauDEM/TauDEM-Develop/src/tiffIO.h:49,
                 from /home/svimal/Software/TauDEM/TauDEM-Develop/src/tiffIO.cpp:47:
/home/svimal/Software/TauDEM/TauDEM-Develop/src/linearpart.h:524:13: error: invalid use of incomplete type ‘class tiffIO’
   dyc[i]=obj.getdyc(globalY);
             ^
In file included from /home/svimal/Software/TauDEM/TauDEM-Develop/src/linearpart.h:42:0,
                 from /home/svimal/Software/TauDEM/TauDEM-Develop/src/commonLib.h:103,
                 from /home/svimal/Software/TauDEM/TauDEM-Develop/src/tiffIO.h:49,
                 from /home/svimal/Software/TauDEM/TauDEM-Develop/src/tiffIO.cpp:47:
/home/svimal/Software/TauDEM/TauDEM-Develop/src/partition.h:46:7: note: forward declaration of ‘class tiffIO’
 class tiffIO;
       ^
CMakeFiles/streamnet.dir/build.make:134: recipe for target 'CMakeFiles/streamnet.dir/tiffIO.cpp.o' failed
make[2]: *** [CMakeFiles/streamnet.dir/tiffIO.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/streamnet.dir/all' failed
make[1]: *** [CMakeFiles/streamnet.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
`
PS: Maybe I should just start using TauDEM through Hydroshare's Jupyter Notebooks?

**Comment by eculler** [comment]
I had a similar problem compiling on MacOS 10.13.2, with a more detailed error message:

 [  3%] Building CXX object CMakeFiles/threshold.dir/tiffIO.cpp.o
In file included from /Users/elsaculler/Downloads/TauDEM-Develop/src/tiffIO.cpp:47:
In file included from /Users/elsaculler/Downloads/TauDEM-Develop/src/tiffIO.h:49:
In file included from /Users/elsaculler/Downloads/TauDEM-Develop/src/commonLib.h:103:
/Users/elsaculler/Downloads/TauDEM-Develop/src/linearpart.h:523:16: error: 
      member access into incomplete type 'tiffIO'
            dxc[i]=obj.getdxc(globalY);
                      ^
/Users/elsaculler/Downloads/TauDEM-Develop/src/partition.h:46:7: note: forward
      declaration of 'tiffIO'
class tiffIO;
      ^
In file included from /Users/elsaculler/Downloads/TauDEM-Develop/src/tiffIO.cpp:47:
In file included from /Users/elsaculler/Downloads/TauDEM-Develop/src/tiffIO.h:49:
In file included from /Users/elsaculler/Downloads/TauDEM-Develop/src/commonLib.h:103:
/Users/elsaculler/Downloads/TauDEM-Develop/src/linearpart.h:524:13: error: 
      member access into incomplete type 'tiffIO'
                dyc[i]=obj.getdyc(globalY);
                          ^
/Users/elsaculler/Downloads/TauDEM-Develop/src/partition.h:46:7: note: forward
      declaration of 'tiffIO'
class tiffIO;
      ^
/Users/elsaculler/Downloads/TauDEM-Develop/src/tiffIO.cpp:247:2: warning: 
      ignoring return value of function declared with 'warn_unused_result'
      attribute [-Wunused-result]
        GDALRasterIO(bandh, GF_Read, xstart, ystart, numCols, numRows,
        ^~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/elsaculler/Downloads/TauDEM-Develop/src/tiffIO.cpp:366:3: warning: 
      ignoring return value of function declared with 'warn_unused_result'
      attribute [-Wunused-result]
                GDALRasterIO(bandh, GF_Write, xstart, ystart, numCols, numRows,
                ^~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/elsaculler/Downloads/TauDEM-Develop/src/tiffIO.cpp:404:3: warning: 
      ignoring return value of function declared with 'warn_unused_result'
      attribute [-Wunused-result]
                GDALRasterIO(bandh, GF_Write, xstart, ystart, numCols, numRows,
                ^~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3 warnings and 2 errors generated.
make[2]: *** [CMakeFiles/threshold.dir/tiffIO.cpp.o] Error 1
make[1]: *** [CMakeFiles/threshold.dir/all] Error 2
make: *** [all] Error 2

My problem was solved by downloading the 5.3.8 tag rather than the develop branch.

**Comment by solomonvimal** [comment]
This solved it for me! Thanks!!

Solomon Vimal
